src/displayapp/screens/Twos.cpp

```diff
@@ -97,39 +92,37 @@ Twos::~Twos() {
 }
 
 bool Twos::placeNewTile() {
-  std::vector<std::pair<int, int>> availableCells;
-  for (int row = 0; row < 4; row++) {
-    for (int col = 0; col < 4; col++) {
-      if (!grid[row][col].value) {
-        availableCells.push_back(std::make_pair(row, col));
-      }
+  unsigned int emptyCells[nCells];
+  unsigned int nEmpty = 0;
+  for (unsigned int i = 0; i < nCells; i++) {
+    const unsigned int row = i / nCols;
+    const unsigned int col = i % nCols;
+    if (grid[row][col].value == 0) {
+      emptyCells[nEmpty] = i;
+      nEmpty++;
     }
   }
 
-  if (availableCells.size() == 0) {
+  if (nEmpty == 0) {
     return false; // game lost
   }
 
-  auto it = availableCells.cbegin();
-  int random = rand() % availableCells.size();
-  std::advance(it, random);
-  std::pair<int, int> newCell = *it;
+  int random = rand() % nEmpty;
 
-  if ((rand() % 100) < 90)
-    grid[newCell.first][newCell.second].value = 2;
-  else
-    grid[newCell.first][newCell.second].value = 4;
-  updateGridDisplay(grid);
+  if ((rand() % 100) < 90) {
+    grid[emptyCells[random] / nCols][emptyCells[random] % nCols].value = 2;
+  } else {
+    grid[emptyCells[random] / nCols][emptyCells[random] % nCols].value = 4;
+  }
+  updateGridDisplay();
   return true;
 }
 
-bool Twos::tryMerge(TwosTile grid[][4], int& newRow, int& newCol, int oldRow, int oldCol) {
+bool Twos::tryMerge(int newRow, int newCol, int oldRow, int oldCol) {
   if (grid[newRow][newCol].value == grid[oldRow][oldCol].value) {
     if ((newCol != oldCol) || (newRow != oldRow)) {
       if (!grid[newRow][newCol].merged) {
-        unsigned int newVal = grid[oldRow][oldCol].value *= 2;
-        grid[newRow][newCol].value = newVal;
-        score += newVal;
+        score += grid[newRow][newCol].value *= 2;
```

Comment:
I think this statement is a bit too optimized, can we split it up again?

```suggestion
        grid[newRow][newCol].value *= 2;
        score += grid[newRow][newCol].value;
```